File: app/src/main/java/com/aspire/dubsmash/siavash/FragmentSounds.java
package com.aspire.dubsmash.siavash;

import android.app.Fragment;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Switch;

import com.aspire.dubsmash.R;
import com.aspire.dubsmash.util.Constants;
import com.aspire.dubsmash.util.Util;

import java.util.ArrayList;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

/**
 * Created by sia on 9/30/15.
 */
public class FragmentSounds extends Fragment {
    private static final String TAG = FragmentSounds.class.getSimpleName();

    @Bind(R.id.switch_lt) Switch mSwitchLT;
    @Bind(R.id.items_recycler_view) RecyclerView mSoundsRecyclerView;

    private AdapterSounds mAdapter;
    private List<Sound> mSounds;
    private boolean isViewShown = false;
    private boolean isFirstRun = true;
    private int group = 0, quantity = 15;

    @Nullable @Override public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_sounds_videos, container, false);
        ButterKnife.bind(this, view);
        mSounds = new ArrayList<>();
        setUpSwitch();
        if (isFirstRun) downloadInitialSounds();
        return view;
    }

    @Override public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isVisibleToUser && getView() != null) {
            downloadInitialSounds();
        }
    }

    private void setUpSwitch() {
        mSwitchLT.setTextOff(Constants.TRENDING);
        mSwitchLT.setTextOn(Constants.LATEST);
        mSwitchLT.setChecked(true);
    }

    private void setUpRecyclerView() {
        isFirstRun = false;
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActivity());
        mSoundsRecyclerView.setLayoutManager(linearLayoutManager);
        mAdapter = new AdapterSounds(mSounds);
        mSoundsRecyclerView.setAdapter(mAdapter);
        mSoundsRecyclerView.addItemDecoration(new SpacesItemDecoration(Constants.RECYCLE_VIEW_SPACE));
        mSoundsRecyclerView.addOnScrollListener(new EndlessRecyclerOnScrollListener<LinearLayoutManager>(linearLayoutManager) {
            @Override public void onLoadMore(int currentPage) {
                if (mSwitchLT.isChecked()) {
                    downloadSounds(Constants.LATEST, String.valueOf(++group), String.valueOf(quantity));
                } else {
                    downloadSounds(Constants.TRENDING, String.valueOf(++group), String.valueOf(quantity));
                }
            }
        });
    }

    private void downloadSounds(final String retrieveMode, String group, String quantity) {
        ActivityMain.sNetworkApi.downloadSound(retrieveMode, group, quantity, null, new Callback<MultipleResult<Sound>>() {
            @Override public void success(MultipleResult<Sound> soundMultipleResult, Response response) {
                if (soundMultipleResult.getOk().equals(Constants.OK)) {
                    mSounds.addAll(soundMultipleResult.getItems());
                    if (isFirstRun) setUpRecyclerView();
                    else mAdapter.addData(mSounds);
                }
            }

            @Override public void failure(RetrofitError error) {
            }
        });
    }

    private void downloadInitialSounds() {
        if (mSwitchLT.isChecked()) {
            downloadSounds(Constants.LATEST, String.valueOf(group), String.valueOf(quantity));
        } else {
            downloadSounds(Constants.TRENDING, String.valueOf(group), String.valueOf(quantity));
        }
    }

    @Override public void onDetach() {
        super.onDetach();
        if (mAdapter != null && mAdapter.getMediaPlayer() != null) {
            Util.stopAndReleaseMediaPlayer(mAdapter.getMediaPlayer());
        }
    }
}

File: app/src/main/kotlin/com.aspire.dubsmash/fragments/FragmentSounds.kt
package com.aspire.dubsmash.fragments

import android.app.Fragment
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Switch
import com.aspire.dubsmash.EndlessRecyclerOnScrollListener
import com.aspire.dubsmash.R
import com.aspire.dubsmash.adapters.AdapterSounds
import com.aspire.dubsmash.util.*
import org.jetbrains.anko.ctx
import retrofit.Callback
import retrofit.Response
import retrofit.Retrofit

/**
 * Created by sia on 11/20/15.
 */
class FragmentSounds : Fragment() {
    private val switchLT: Switch by bindView(R.id.switch_lt)
    private val soundsRecyclerView: RecyclerView by bindView(R.id.items_recycler_view)

    private var adapter: AdapterSounds? = null
    private var sounds: MutableList<Sound> = arrayListOf()
    private var isFirstRun = true
    private var group = 0
    private val quantity = 15

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_sounds_videos, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setUpSwitch()
        if (isFirstRun) downloadInitialSounds()
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (isVisibleToUser && view != null) downloadInitialSounds()
    }

    private fun setUpSwitch() {
        switchLT.textOff = trending
        switchLT.textOn = latest
        switchLT.isChecked = true
    }

    private fun setUpRecyclerView() {
        isFirstRun = false
        val linearLayoutManager = LinearLayoutManager(ctx)
        soundsRecyclerView.layoutManager = linearLayoutManager
        adapter = AdapterSounds(sounds)
        soundsRecyclerView.adapter = adapter
        soundsRecyclerView.addOnScrollListener(object : EndlessRecyclerOnScrollListener<LinearLayoutManager>(linearLayoutManager) {
            override fun onLoadMore(currentPage: Int) {
                if (switchLT.isChecked)
                    downloadSounds(latest, (++group).toString(), quantity.toString())
                else
                    downloadSounds(trending, (++group).toString(), quantity.toString())
            }
        })
    }

    private fun downloadSounds(retrieveMode: String, group: String, quantity: String) {
        val result = networkApi.downloadSounds(retrieveMode, group, quantity)
        result.enqueue(object : Callback<MultipleResult<Sound>> {
            override fun onResponse(response: Response<MultipleResult<Sound>>, retrofit: Retrofit?) {
                if (response.isSuccess && response.body().ok == "ok") {
                    sounds.addAll(response.body().items)
                    if (isFirstRun) setUpRecyclerView() else adapter?.addData(sounds)
                }
            }

            override fun onFailure(t: Throwable?) {
                throw UnsupportedOperationException()
            }

        })
    }

    private fun downloadInitialSounds() {
        if (switchLT.isChecked)
            downloadSounds(latest, group.toString(), quantity.toString())
        else
            downloadSounds(trending, group.toString(), quantity.toString())

    }

    override fun onDetach() {
        super.onDetach()
        stopAndReleaseMediaPlayer(adapter?.mediaPlayer)
    }
}